federated_aws_cli/login.py

```diff
@@ -3,111 +3,176 @@
+        :param client_id: OIDC client_id of the native OIDC application
+        :param scope: OIDC scopes of claims to request
+        """
+        self.well_known_url = well_known_url
+        self.client_id = client_id
+        self.scope = scope
+        self.tokens = None
+        self.port = port if port is not None else listener.get_available_port()
+        self.redirect_uri = "http://localhost:{}/redirect_uri".format(self.port)
+
+    def __deferred_init__(self):
+        self.openid_configuration = requests.get(self.well_known_url).json()
+        self.jwks = requests.get(self.openid_configuration["jwks_uri"]).json()
+        self.authorization_endpoint = self.openid_configuration["authorization_endpoint"]
+        self.token_endpoint = self.openid_configuration["token_endpoint"]
+        self.code_verifier = self.base64_without_padding(os.urandom(32))
+        self.code_challenge = self.generate_challenge(self.code_verifier)
+        self.state = self.base64_without_padding(os.urandom(32))
+
+    def base64_without_padding(self, data):
+        # https://tools.ietf.org/html/rfc7636#appendix-A
+        return base64.urlsafe_b64encode(data).decode("utf-8").rstrip("=")
+
+    def generate_challenge(self, code_verifier):
+        # https://tools.ietf.org/html/rfc7636#section-4.2
+        return self.base64_without_padding(hashlib.sha256(code_verifier.encode()).digest())
+
+    def as_env_variables(self):
+        """
+        :return: str of environment
+        """
+        verb = "set" if platform.system() == "Windows" else "export"
+        data = "{} ID_TOKEN={}\n".format(verb, self.tokens["id_token"])
+        data += '{} ID_TOKEN_EXPIRES_AT="{}"'.format(verb, self.tokens["expires_at"])
+        return data
+
+    def refresh_id_token(self, laytime=120):
+        """
+        Refresh credentials as necessary if expired
+        :laytime: int: How many seconds of laytime between requests to get new credentials, when approaching the
+        expiration window
+        """
+        self.__deferred_init__()
+        if self.tokens is None:
+            tmp_token = os.environ.get("ID_TOKEN")
+            tmp_exp = os.environ.get("ID_TOKEN_EXPIRES_AT")
+            if tmp_token is not None and tmp_exp is not None:
+                self.tokens = {"id_token": tmp_token, "expires_at": tmp_exp}
+                logger.debug("Loaded existing id_token from the environment")
+            else:
+                logger.debug("We have no id_token, getting a new one")
+                return self.get_wait_for_id_token()
+
+        # Instruct we base ourselves on UTC
+        os.environ["TZ"] = "UTC"
+        try:
+            now = datetime.now(timezone.utc)
+        except NameError:
+            # P2
+            now = datetime.now()
+        exp = parser.parse(self.tokens["expires_at"])
+        diff = exp - now
+        # Make this an int with seconds since the epoch
```

Comment:
Use `expiration` instead of shorthand?

Also, no need to assign a variable used once, you can just do:
`diff = parser.parse(self.tokens["expires_at"]) - now`

Reply:
this is mainly for clarity, but am happy to go either way